### PR TITLE
ToggleGroupControl: Don't set value on focus after a reset

### PR DESCRIPTION
**Pull request by mirka**
Bug discovered in https://github.com/WordPress/gutenberg/pull/65386#discussion_r1798762479

## What?

Fixes an incomplete bug fix in #65892 where ToggleGroupControl could still autoselect an option on focus after the selected value was reset by the consumer.

### Testing Instructions for Keyboard

Modify the Storybook story so it has a reset button, like this:

```diff
diff --git a/packages/components/src/toggle-group-control/stories/index.story.tsx b/packages/components/src/toggle-group-control/stories/index.story.tsx
index 92f1e60762..01e5fc4eae 100644
--- a/packages/components/src/toggle-group-control/stories/index.story.tsx
+++ b/packages/components/src/toggle-group-control/stories/index.story.tsx
@@ -48,15 +48,18 @@ const Template: StoryFn< typeof ToggleGroupControl > = ( {
 		useState< ToggleGroupControlProps[ 'value' ] >();
 
 	return (
-		<ToggleGroupControl
-			__nextHasNoMarginBottom
-			{ ...props }
-			onChange={ ( ...changeArgs ) => {
-				setValue( ...changeArgs );
-				onChange?.( ...changeArgs );
-			} }
-			value={ value }
-		/>
+		<>
+			<ToggleGroupControl
+				__nextHasNoMarginBottom
+				{ ...props }
+				onChange={ ( ...changeArgs ) => {
+					setValue( ...changeArgs );
+					onChange?.( ...changeArgs );
+				} }
+				value={ value }
+			/>
+			<button onClick={ () => setValue( undefined ) }>Reset</button>
+		</>
 	);
 };

1. Select a value.
2. Reset the value using the reset button.
3. Using the keyboard, focus back onto the component. A value should not be selected automatically.

```

### CODE REVIEW

File: packages/components/src/toggle-group-control/toggle-group-control-option-base/component.tsx [hunk 1]
```diff
@@ -141,10 +141,14 @@ function ToggleGroupControlOptionBase(
 					<Ariakit.Radio
 						disabled={ disabled }
 						onFocusVisible={ () => {
+							const selectedValueIsNullish =
+								toggleGroupControlContext.value === null ||
+								toggleGroupControlContext.value === '';
```
Comment (mirka) [inline packages/components/src/toggle-group-control/toggle-group-control-option-base/component.tsx:146]: Due to how we add a layer of [`value` conversion](https://github.com/WordPress/gutenberg/blob/26e108deabcb6f58ffdf424f06887efb9f88333c/packages/components/src/toggle-group-control/toggle-group-control/as-radio-group.tsx#L51-L56) in controlled mode, the Ariakit store will hold a `null` value on first render with no value provided, but then an empty string after a consumer resets the `value` to `undefined`.

File: packages/components/src/toggle-group-control/toggle-group-control/as-radio-group.tsx [hunk 2]
```diff
@@ -73,6 +73,13 @@ function UnforwardedToggleGroupControlAsRadioGroup(
 	const selectedValue = useStoreState( radio, 'value' );
 	const setValue = radio.setValue;
 
+	// Ensures that the active id is also reset after the value is "reset" by the consumer.
+	useEffect( () => {
+		if ( selectedValue === '' ) {
+			radio.setActiveId( undefined );
+		}
+	}, [ radio, selectedValue ] );
```
Comment (mirka) [inline packages/components/src/toggle-group-control/toggle-group-control/as-radio-group.tsx:81]: I can feel the tech debt catching up to us 😅

File: packages/components/src/toggle-group-control/toggle-group-control-option-base/component.tsx [hunk 3]
```diff
@@ -141,10 +141,14 @@ function ToggleGroupControlOptionBase(
 					<Ariakit.Radio
 						disabled={ disabled }
 						onFocusVisible={ () => {
+							const selectedValueIsNullish =
```
Comment (tyxla) [inline packages/components/src/toggle-group-control/toggle-group-control-option-base/component.tsx]: Naming nit: `nullish` often refers to `null` or `undefined`, so if someone isn't reading closely, they might be confused and expect this checks for `undefined` and not an empty string.

File: packages/components/src/toggle-group-control/toggle-group-control/as-radio-group.tsx (hunk 2 above)
Comment (tyxla) [reply packages/components/src/toggle-group-control/toggle-group-control/as-radio-group.tsx:81]: Food for thought: I wonder if this can be improved with a centralized controlled component layer that all controlled components would use. Probably a too idealistic idea, since some components use Ariakit and others do not, and this layer would be quite complicated to cover all cases.

File: packages/components/src/toggle-group-control/test/index.tsx [hunk 4]
```diff
@@ -175,6 +175,32 @@ describe.each( [
 		expect( radio ).not.toBeChecked();
 	} );
 
+	if ( mode === 'controlled' ) {
```
Comment (tyxla) [inline packages/components/src/toggle-group-control/test/index.tsx:178]: Should we move it next to the other `controlled`-only tests?

**Review by tyxla — APPROVED** [review]
LGTM, thanks for the fix 👍 🚀

File: packages/components/src/toggle-group-control/toggle-group-control/as-radio-group.tsx (hunk 2 above)
Comment (ciampo) [reply packages/components/src/toggle-group-control/toggle-group-control/as-radio-group.tsx:81]: I think that the most practical fix here is to change the APIs to follow the standard convention for controlled / uncontrolled component. Since that would be a breaking change, we could just release a V2 of the component that would give us also a chance to:

- implement / test [the new design spec](https://github.com/wordPress/gutenberg/issues/64439) alongside the legacy one
- review the current API strategy around "deselectability" and potentially multiple selection (should they be separate components? Or different subcomponents?)

File: packages/components/src/toggle-group-control/toggle-group-control-option-base/component.tsx (hunk 3 above)
Comment (ciampo) [reply packages/components/src/toggle-group-control/toggle-group-control-option-base/component.tsx]: Maybe `selectedValueIsEmpty` ? Or even inlining the check?

Comment (tyxla) [reply packages/components/src/toggle-group-control/toggle-group-control-option-base/component.tsx]: I'd be fine with inlining it 👍

File: packages/components/src/toggle-group-control/toggle-group-control/as-radio-group.tsx (hunk 2 above)
Comment (mirka) [reply packages/components/src/toggle-group-control/toggle-group-control/as-radio-group.tsx:81]: For this case specifically, I think there might be room to argue that resetting the active id on empty-ish (or invalid) values should be handled by Ariakit out of the box. I'll try and follow up.

File: packages/components/src/toggle-group-control/toggle-group-control-option-base/component.tsx (hunk 3 above)
Comment (mirka) [reply packages/components/src/toggle-group-control/toggle-group-control-option-base/component.tsx]: Good point, I'll go with `selectedValueIsEmpty` here. (It started as an inline condition, but it was hard to parse/understand so I prefer it named.)

File: packages/components/src/toggle-group-control/test/index.tsx (hunk 4 above)
Comment (mirka) [reply packages/components/src/toggle-group-control/test/index.tsx:178]: No strong opinion, but I think I prefer it here because it's closely related to the test right above it.

**Comment by ciampo** [comment]
@mirka just a ping in case you didn't notice the failed backport

**Comment by mirka** [comment]
Backported at 4a7de72e65f54a13a315492abd2c52988264f514
